include/net/net_pkt.h

```diff
@@ -1998,6 +2001,158 @@ const char *net_pkt_pool2str(struct net_buf_pool *pool);
+	net_pkt_alloc_buffer_debug(_pkt, _size, _proto, _timeout,	\
+				      __func__, __LINE__)
+
+struct net_pkt *net_pkt_alloc_with_buffer_debug(struct net_if *iface,
+						size_t size,
+						sa_family_t family,
+						enum net_ip_protocol proto,
+						s32_t timeout,
+						const char *caller,
+						int line);
+#define net_pkt_alloc_with_buffer(_iface, _size, _family,		\
+				  _proto, _timeout)			\
+	net_pkt_alloc_with_buffer_debug(_iface, _size, _family,	\
+					   _proto, _timeout,		\
+					   __func__, __LINE__)
+
+struct net_pkt *net_pkt_rx_alloc_with_buffer_debug(struct net_if *iface,
+						   size_t size,
+						   sa_family_t family,
+						   enum net_ip_protocol proto,
+						   s32_t timeout,
+						   const char *caller,
+						   int line);
+#define net_pkt_rx_alloc_with_buffer(_iface, _size, _family,		\
+				     _proto, _timeout)			\
+	net_pkt_rx_alloc_with_buffer_debug(_iface, _size, _family,	\
+					      _proto, _timeout,		\
+					      __func__, __LINE__)
+#else
+
+/**
+ * @brief Allocate an initialized net_pkt
+ *
+ * Note: for the time being, 2 pools are used. One for TX and one for RX.
+ *       This allocater has to be used for TX.
+ *
+ * @param timeout Maximum time in milliseconds to wait for an allocation.
+ *
+ * @return a pointer to a newly allocated net_pkt on success, NULL otherwise.
+ */
+struct net_pkt *net_pkt_alloc(s32_t timeout);
+
+/**
+ * @brief Allocate an initialized net_pkt for RX
+ *
+ * Note: for the time being, 2 pools are used. One for TX and one for RX.
+ *       This allocater has to be used for RX.
+ *
+ * @param timeout Maximum time in milliseconds to wait for an allocation.
+ *
+ * @return a pointer to a newly allocated net_pkt on success, NULL otherwise.
+ */
+struct net_pkt *net_pkt_rx_alloc(s32_t timeout);
+
+/**
+ * @brief Allocate a network packet for a specific network interface.
+ *
+ * @param iface The network interface the packet is supposed to go through.
+ * @param timeout Maximum time in milliseconds to wait for an allocation.
+ *
+ * @return a pointer to a newly allocated net_pkt on success, NULL otherwise.
+ */
+struct net_pkt *net_pkt_allocate_on_iface(struct net_if *iface, s32_t timeout);
+
+/* Same as above but specifically for RX packet */
+struct net_pkt *net_pkt_rx_alloc_on_iface(struct net_if *iface, s32_t timeout);
+
+/**
+ * @brief Allocate buffer for a net_pkt
+ *
+ * Note: such allocator will take into account space necessary for headers,
+ *       MTU, and existing buffer (if any). Beware that, due to all these
+ *       criterias, the allocated size might be smaller/bigger than requested
+ *       one.
+ *
+ * @param pkt     The network packet requiring buffer to be allocated.
+ * @param size    The size of buffer being requested.
+ * @param proto   The IP protocol type (can be 0 for none).
+ * @param timeout Maximum time in milliseconds to wait for an allocation.
+ *
+ * @return 0 on success, negative errno code otherwise.
```

Comment:
nd - how to access this new buffer allocated?

Reply:
It's still allocated through the infamous frags attribute (I really don't like this name, it's mixes up fragment of memory vs fragment of a packet, which is a concept that is present in IP also... meh, I introduced a "buffer" attribute, in a union with frags. Which one will disappear). So current API works the same. But the new API introduced in next patch is far easier to use. (one does not have to care about looping on "frags", having its own offset, frag, etc...).